src/draw.c

```diff
@@ -742,26 +742,48 @@ static void render_content(cairo_t *c, struct colored_layout *cl, int width, dou
 
                 // draw progress bar
                 // Note: the bar could be drawn a bit smaller, because the frame is drawn on top
-                // left side
+                // left side (fill)
                 cairo_set_source_rgba(c, cl->highlight.r, cl->highlight.g, cl->highlight.b, cl->highlight.a);
-                draw_rect(c, x_bar_1, frame_y, progress_width_1, progress_height, scale);
+                if(settings.progress_bar_corner_radius > 0) {
+                        draw_rounded_rect(c, x_bar_1, frame_y, progress_width_1, progress_height, 
+                                settings.progress_bar_corner_radius, scale, true, true);
+                } else {
+                        draw_rect(c, x_bar_1, frame_y, progress_width_1, progress_height, scale);
+                }
                 cairo_fill(c);
-                // right side
+                // right side (background)
                 cairo_set_source_rgba(c, cl->bg.r, cl->bg.g, cl->bg.b, cl->bg.a);
-                draw_rect(c, x_bar_2, frame_y, progress_width_2, progress_height, scale);
+                if(settings.progress_bar_corner_radius > 0) {
+                        draw_rounded_rect(c, x_bar_2, frame_y, progress_width_2, progress_height, 
+                                settings.progress_bar_corner_radius, scale, true, true);
+                } else {
+                        draw_rect(c, x_bar_2, frame_y, progress_width_2, progress_height, scale);
+                }
                 cairo_fill(c);
+
                 // border
                 cairo_set_source_rgba(c, cl->frame.r, cl->frame.g, cl->frame.b, cl->frame.a);
-                // TODO draw_rect instead of cairo_rectangle resulted
-                // in blurry lines due to rounding (half_frame_width
-                // can be non-integer)
-                cairo_rectangle(c,
-                                (frame_x + half_frame_width) * scale,
-                                (frame_y + half_frame_width) * scale,
-                                (progress_width - frame_width) * scale,
-                                progress_height * scale);
                 cairo_set_line_width(c, frame_width * scale);
-                cairo_stroke(c);
+                if(settings.progress_bar_corner_radius > 0) {
+                        draw_rounded_rect(c,
+                                        (frame_x + half_frame_width),
+                                        (frame_y + half_frame_width),
+                                        (progress_width - frame_width),
+                                        progress_height,
+                                        settings.progress_bar_corner_radius,
+                                        scale, true, true);
+                        cairo_stroke(c);
+                } else {
+                        // TODO draw_rect instead of cairo_rectangle resulted
```

Comment:
This TODO explains it a bit, but for the drawing of the rectangle, you need a non-integer position. The position gives the center of the line. So if you want a horizontal line of 1 pixel width at height = 100 to not be blurry, you need to give a y-coordinate of 100.5.
I think to solve this, you will need to change the position arguments of draw_rounded_rect to floats.

Reply:
Thanks for all of the detailed feedback. I've pushed a commit which addresses every comment except this one. How should the subpixel position be calculated? `x = (x * scale)` rounded to the nearest .5? Do we add or subtract .5 if starting from an integer? Thank you

Reply:
Yeah that's a tough one. If you never want blurry lines, you would have to do some logic like that. Since we don't officially support fractional scaling, you might leave that as a TODO (make a comment in that case). With whole number scaling it should work like this.
If you want to support fractional scaling, I think the calculation should be done inside `draw_rounded_rect`. It needs the additional information of the line width. Then the calculation is as follows:
`x = round(x*scale) + half_frame_width`